### PR TITLE
Types for P6 GenesisData

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Introduces changes needed for genesis-creator to support P6
Related to https://github.com/Concordium/concordium-misc-tools/pull/51

## Changes

- Introduce `ChainParametersV2` struct for protocol version 6.
- Introduce generic `gas_reward` in `RewardParametersSkeleton` for supporting different versions of GASRewards.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/types/mod.rs [hunk 1]
```diff
@@ -1792,6 +1792,33 @@ pub struct ChainParametersV1 {
     pub pool_parameters:          PoolParameters,
 }
 
+#[derive(common::Serial, Debug)]
```
Comment (abizjak) [inline src/types/mod.rs:1795]: We should have Deserial as well I think. Unless there is a reason not to?

File: src/types/mod.rs [hunk 2]
```diff
@@ -1804,27 +1831,33 @@ impl ChainParametersFamily for ChainParameterVersion1 {
...
+    type Output = ChainParametersV2;
+}
+
 pub type ChainParameters<CPV> = <CPV as ChainParametersFamily>::Output;
 
 #[derive(Debug, SerdeSerialize, SerdeDeserialize)]
 #[serde(rename_all = "camelCase")]
 /// Values of reward parameters.
-pub struct RewardParametersSkeleton<MD> {
+pub struct RewardParametersSkeleton<MD, GR> {
```
Comment (abizjak) [inline src/types/mod.rs:1843]: It'd be useful to comment on what GR and MD are so we can better maintain this.

Alternatively name the type parameters in a more verbose way.

File: src/types/mod.rs [hunk 3]
```diff
@@ -1792,6 +1792,33 @@ pub struct ChainParametersV1 {
     pub pool_parameters:          PoolParameters,
 }
 
+#[derive(common::Serial, Debug)]
+/// Values of chain parameters that can be updated via chain updates.
+pub struct ChainParametersV2 {
+    /// Consensus protocol version 2 timeout parameters.
+    pub timeout_parameters:                TimeoutParameters,
+    /// Minimum time interval between blocks.
+    pub min_block_time:                    SlotDuration,
```
Comment (abizjak) [inline src/types/mod.rs:1801]: Is there a reason why we don't just use Duration here?

File: src/types/mod.rs [hunk 4]
```diff
@@ -1792,6 +1792,33 @@ pub struct ChainParametersV1 {
...
+    /// Maximum energy allowed per block.
+    pub block_energy_limit:                Energy,
+    /// Euro per energy exchange rate.
+    pub euro_per_energy:                   ExchangeRate,
+    /// Micro ccd per euro exchange rate.
+    pub micro_gtu_per_euro:                ExchangeRate,
```
Comment (abizjak) [inline src/types/mod.rs:1807]: Maybe call it `micro_ccd_per_euro` now? :)

**Review by abizjak — APPROVED** [review]

File: src/types/mod.rs (hunk 3 above)
Comment (limemloh) [reply src/types/mod.rs:1801]: Right, when introducing this type, Duration did not implement `concordium_base::Serial`.
I'll change this

File: src/types/mod.rs (hunk 2 above)
Comment (limemloh) [reply src/types/mod.rs]: I have made the names of the generics more explicit and added a bit to the documentation

**Pull request merged**